teps/0004-tekton-integrations-annotations-statuses.md

```diff
@@ -0,0 +1,689 @@
+
+This relates heavily to existing work around
+[notifications](https://github.com/tektoncd/pipeline/issues/1740), and aims to
+build on it to support metadata for arbitrary integrations alongside the
+existing
+[Cloud Event notifications](https://github.com/tektoncd/pipeline/issues/2082).
+While this proposal is largely compatible with existing work, one change would
+be recommended - notably migrating `CloudEventDelivery` as an integration
+status.
+
+Cloud Events are effectively a common type of a specific integration, and would
+fit well into this model in order to support arbitrary integration data. The
+Pipeline controller would be able to continue to insert data just like any other
+integration, and would be able to continue to write existing CloudEventDelivery
+status data alongside IntegrationStatus data.
+
+Example:
+
+[CloudEventDelivery](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1?tab=doc#CloudEventDelivery):
+
+```yaml
+target: example.com
+status:
+  condition: Failed
+  sentAt: 123456789
+  error: "access denied"
+  retryCount: 1
+```
+
+IntegrationStatus:
+
+```yaml
+- name: cloudevent.integrations.tekton.dev
+  conditions:
+    - type: Succeeded
+      status: False
+      reason: "access denied"
+  lastTransitionTime: 123456789
+  annotations:
+    target: example.com
+    retryCount: 1
+```
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+#### External Integration Quotas
+
+If relying on reconcilers / events that trigger on every update, integration
+providers will need to be careful to stay within quotas external providers may
+impose.
+
+While Tekton Pipelines will not impose specific throttling for integration
+providers, it will be a general best practice to use Status/Condition data (e.g.
+`ObservedGeneration`, `Conditions`, `LastTransitionTime`) as well as reconciler
+push back with `RequeueAfter` to stay within external quotas.
+
+### User Experience (optional)
+
+<!--
+Consideration about the user experience. Depending on the area of change,
+users may be task and pipeline editors, they may trigger task and pipeline
+runs or they may be responsible for monitoring the execution of runs,
+via CLI, dashboard or a monitoring system.
+
+Consider including folks that also work on CLI and dashboard.
+-->
+
+#### Storing data in Pipelines
+
+Similar to `TaskRunResults` and `PipelineResourceResults`, integration statuses
+are another form of external artifact generated as a result of the Pipeline.
+Because of this, it makes sense to store these statuses as a part of the
+Pipeline/TaskRun status, so that it is easy to find and see this data when
+something goes wrong.
```

Comment:
I am not entirely sure about this. `TaskRunResults` and `PipelineResourceResults` are, by definition, results that are coming from the execution of the Task/Pipeline (aka what the task is supposed to do, like clone a git repository, …) through TaskRun/PipelineRun) This would not be the case with integration, as they do not necessary result from the execution of it (aka what the task/pipeline does), but only if it executed or not, …